README.rst
========
Treepoem
========

.. image:: https://img.shields.io/github/actions/workflow/status/adamchainz/treepoem/main.yml?branch=main&style=for-the-badge
   :target: https://github.com/adamchainz/treepoem/actions?workflow=CI

.. image:: https://img.shields.io/pypi/v/treepoem.svg?style=for-the-badge
   :target: https://pypi.org/project/treepoem/

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg?style=for-the-badge
   :target: https://github.com/psf/black

.. image:: https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white&style=for-the-badge
   :target: https://github.com/pre-commit/pre-commit
   :alt: pre-commit

A cleverly named, but very simple python barcode renderer wrapping the
BWIPP_ library and ``ghostscript`` command line tool.

Installation
============

Install from **pip**:

.. code-block:: sh

    python -m pip install treepoem

Python 3.8 to 3.12 supported.

You'll also need Ghostscript installed. On Ubuntu/Debian this can be installed
with:

.. code-block:: sh

    apt-get install ghostscript

On Mac OS X use:

.. code-block:: sh

    brew install ghostscript

Otherwise refer to your distribution's package manager, though it's likely to
be called ``ghostscript`` too.

There's a known issue with rendering on Ghostscript 9.22+ where images are
smeared. See
`GitHub Issue #124 <https://github.com/adamchainz/treepoem/issues/124>`_ and
its associated links for more details. Ghostscript merged a fix in version
9.26 and common barcodes seem to work from then on, though still with some
smearing.

You can check your Ghostscript version with:

.. code-block:: sh

    gs --version

----

**Working on a Django project?**
Improve your skills with `one of my books <https://adamj.eu/books/>`__.

----

API
===

``generate_barcode(barcode_type: str, data: str | bytes, options: dict[str, str | bool] | None=None, scale: int = 2) -> EpsImageFile``
----------------------------------------------------------------------------------------------------------------------

Generates a barcode and returns it as a PIL image file object (specifically, a
``PIL.EpsImagePlugin.EpsImageFile``).

``barcode_type`` is the name of the barcode type to generate (see below).

``data`` is a ``str`` or ``bytes`` of data to embed in the barcode - the amount
that can be embedded varies by type.

``options`` is a dictionary of strings-to-strings of extra options to be passed
to BWIPP_, as per its docs.

For example, this generates a QR code image, and saves it to a file using
standard PIL ``Image.save()``:

.. code-block:: pycon

   >>> import treepoem
   >>> image = treepoem.generate_barcode(
   ...     barcode_type="qrcode",  # One of the BWIPP supported codes.
   ...     data="barcode payload",
   ... )
   >>> image.convert("1").save("barcode.png")

If your barcode image is monochrome, with no additional text or
coloring, converting the ``Image`` object to monochrome as shown above
(``image.convert('1')``) will likely reduce its file size.

``barcode_types: dict[str, BarcodeType]``
-----------------------------------------

This is a ``dict`` of the ~100 names of the barcode types that the vendored
version of BWIPP_ supports: its keys are ``str``\s of the barcode type encoder
names, and the values are instances of ``BarcodeType``.

``BarcodeType``
---------------

A class representing meta information on the types. It has two attributes:

* ``type_code`` - the value needed for the ``barcode_type`` argument of
  ``generate_barcode()`` to use this type.

* ``description`` - the human level description of the type
  which has two ``str``.

Only these common types are used in the test suite:

* ``qrcode`` - `QR Code`_

* ``azteccode`` - `Aztec Code`_

* ``pdf417`` - PDF417_

* ``interleaved2of5`` - `Interleaved 2 of 5`_

* ``code128`` - `Code 128`_

* ``code39`` - `Code 39`_

Command-line interface
======================

Treepoem also includes a simple command-line interface to the
functionality of ``generate_barcode``. For example, these commands
will generate two QR codes with identical contents, but different levels
of error correction (see `QR Code Options`_):

.. code-block:: sh

   $ treepoem -o barcode1.png -t qrcode "This is a test" eclevel=H
   $ treepoem -o barcode2.png -t qrcode "^084his is a test" eclevel=L parse

Complete usage instructions are shown with ``treepoem --help``.

What's so clever about the name?
================================

Barcode.

Bark ode.

Tree poem.

Updating BWIPP
==============

For development of treepoem, when there's a new BWIPP release:

1. Run ``./download_bwipp.py`` with the version of BWIPP to download.
2. Run ``./make_data.py`` to update the barcode types that treepoem knows about.
3. Add a note in ``CHANGELOG.rst`` about the upgrade, adapting from the previous one.
4. Commit and make a pull request, `adapting from previous examples <https://github.com/adamchainz/treepoem/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+upgrade+bwipp>`__.

.. _BWIPP: https://github.com/bwipp/postscriptbarcode
.. _QR Code: https://github.com/bwipp/postscriptbarcode/wiki/QR-Code
.. _Aztec Code: https://github.com/bwipp/postscriptbarcode/wiki/Aztec-Code
.. _PDF417: https://github.com/bwipp/postscriptbarcode/wiki/PDF417
.. _Interleaved 2 of 5: https://github.com/bwipp/postscriptbarcode/wiki/Interleaved-2-of-5
.. _Code 128: https://github.com/bwipp/postscriptbarcode/wiki/Code-128
.. _Code 39: https://github.com/bwipp/postscriptbarcode/wiki/Code-39
.. _QR Code Options: https://github.com/bwipp/postscriptbarcode/wiki/QR-Code


HISTORY.rst
=======
History
=======

* Upgrade vendored BWIPP to its 2022-10-19 release.
  This version includes some more validation improvements and small rendering fixes.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.17.0 (2022-08-18)
-------------------

* Upgrade vendored BWIPP to its 2022-07-29 release.
  This version includes:

  * new helper encoders for GS1 Digital Link URIs included in QR Code and Data Matrix symbols
  * improved validation of GS1 syntax data
  * a friendly escape mechanism for ASCII control characters

  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.16.0 (2022-06-21)
-------------------

* Upgrade vendored BWIPP to its 2022-06-10 release.
  This version includes some fixes for the rendering of certain barcode formats.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.15.0 (2022-05-11)
-------------------

* Support Python 3.11.

3.14.0 (2022-01-10)
-------------------

* Drop Python 3.6 support.

3.13.0 (2021-12-21)
-------------------

* Upgrade vendored BWIPP to its 2021-09-28 release.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.12.0 (2021-09-21)
-------------------

* Upgrade vendored BWIPP to its 2021-07-15 release.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.11.0 (2021-08-13)
-------------------

* Add type hints.

3.10.0 (2021-05-10)
-------------------

* Support Python 3.10.

3.9.0 (2021-03-22)
------------------

* Upgrade vendored BWIPP to its 2021-02-06 release.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

* Stop distributing tests to reduce package size.
  Tests are not intended to be run outside of the tox setup in the repository.
  Repackagers can use GitHub's tarballs per tag.

3.8.0 (2020-12-30)
------------------

* Upgrade vendored BWIPP to its 2020-12-28 release.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.7.0 (2020-12-13)
------------------

* Drop Python 3.5 support.
* Support Python 3.9.

3.6.0 (2020-10-11)
------------------

* Upgrade BWIPP from 2020-09-13 to 2020-10-11.
  This version has a few bug fixes and performance improvements.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.5.0 (2020-09-21)
------------------

* Upgrade BWIPP from 2020-04-01 to 2020-09-13.
  This version has a few bug fixes and performance improvements.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.4.0 (2020-06-21)
------------------

* Upgrade BWIPP from 2019-11-08 to 2020-04-01.
  This version has a few bug fixes and performance improvements.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.3.1 (2020-02-04)
------------------

* Update allowed barcode list to add missing types from new versions of BWIPP.

3.3.0 (2019-12-21)
------------------

* Upgrade BWIPP from 2019-08-05 to 2019-11-08.
  This version has a few bug fixes and performance improvements.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/src/treepoem/postscriptbarcode/CHANGES>`__.

3.2.0 (2019-12-19)
------------------

* Upgrade BWIPP from 2019-04-24 to 2019-08-05.
  This version has a few bug fixes and performance improvements.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/treepoem/postscriptbarcode/CHANGES>`__.

* Converted setuptools metadata to configuration file.
  This meant removing the ``__version__`` attribute from the package.

  If you want to inspect the installed version, use ``importlib.metadata.version("treepoem")`` (`docs <https://docs.python.org/3.8/library/importlib.metadata.html#distribution-versions>`__ / `backport <https://pypi.org/project/importlib-metadata/>`__).

* Update Python support to 3.5-3.8.

3.1.0 (2019-06-25)
------------------

* Update Python support to 3.5-3.7, as 3.4 has reached its end of life.

* Upgrade BWIPP from 2017-07-27 to 2019-04-24.
  This version has a few bug fixes and performance improvements.
  You can read its changelog in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/treepoem/postscriptbarcode/CHANGES>`__.

3.0.0 (2019-05-08)
------------------

* Drop Python 2 support, only Python 3.4+ is supported now.

* Upgrade BWIPP from 2017-05-20 to 2018-07-27.
  This version has a few bug fixes and performance improvements.
  You can read its changelog in the vendored copy in the `treepoem repo <https://github.com/adamchainz/treepoem/blob/main/treepoem/postscriptbarcode/CHANGES>`__.

2.0.0 (2018-08-04)
------------------

* Support binary barcode data - if ``bytes`` (``str`` on Python 2) is passed
  as data, it's not encoded. This has introduced a dependency on ``six``. This
  may be backwards incompatible, depending on what type of data you're passing
  in on Python 2.
* Make ``treepoem.barcode_types`` a ``dict`` mapping the BWIPP encoder
  names to a custom type containing a human-readable ``description``. This is
  backwards incompatible if you're relying on ``barcode_types`` which
  previously was a ``set`` of the encoder names.
* Upgrade BWIPP from 2017-10-19 to 2018-05-20. This has a few bug fixes and

src/treepoem/data.py
from __future__ import annotations


class BarcodeType:
    def __init__(self, type_code: str, description: str) -> None:
        self.type_code = type_code
        self.description = description


# All supported barcode types, extracted from barcode.ps
barcode_types: dict[str, BarcodeType] = {
    "auspost": BarcodeType("auspost", "AusPost 4 State Customer Code"),
    "azteccode": BarcodeType("azteccode", "Aztec Code"),
    "azteccodecompact": BarcodeType("azteccodecompact", "Compact Aztec Code"),
    "aztecrune": BarcodeType("aztecrune", "Aztec Runes"),
    "bc412": BarcodeType("bc412", "BC412"),
    "channelcode": BarcodeType("channelcode", "Channel Code"),
    "codablockf": BarcodeType("codablockf", "Codablock F"),
    "code11": BarcodeType("code11", "Code 11"),
    "code128": BarcodeType("code128", "Code 128"),
    "code16k": BarcodeType("code16k", "Code 16K"),
    "code2of5": BarcodeType("code2of5", "Code 25"),
    "code32": BarcodeType("code32", "Italian Pharmacode"),
    "code39": BarcodeType("code39", "Code 39"),
    "code39ext": BarcodeType("code39ext", "Code 39 Extended"),
    "code49": BarcodeType("code49", "Code 49"),
    "code93": BarcodeType("code93", "Code 93"),
    "code93ext": BarcodeType("code93ext", "Code 93 Extended"),
    "codeone": BarcodeType("codeone", "Code One"),
    "coop2of5": BarcodeType("coop2of5", "COOP 2 of 5"),
    "daft": BarcodeType("daft", "Custom 4 state symbology"),
    "databarexpanded": BarcodeType("databarexpanded", "GS1 DataBar Expanded"),
    "databarexpandedcomposite": BarcodeType(
        "databarexpandedcomposite", "GS1 DataBar Expanded Composite"
    ),
    "databarexpandedstacked": BarcodeType(
        "databarexpandedstacked", "GS1 DataBar Expanded Stacked"
    ),
    "databarexpandedstackedcomposite": BarcodeType(
        "databarexpandedstackedcomposite", "GS1 DataBar Expanded Stacked Composite"
    ),
    "databarlimited": BarcodeType("databarlimited", "GS1 DataBar Limited"),
    "databarlimitedcomposite": BarcodeType(
        "databarlimitedcomposite", "GS1 DataBar Limited Composite"
    ),
    "databaromni": BarcodeType("databaromni", "GS1 DataBar Omnidirectional"),
    "databaromnicomposite": BarcodeType(
        "databaromnicomposite", "GS1 DataBar Omnidirectional Composite"
    ),
    "databarstacked": BarcodeType("databarstacked", "GS1 DataBar Stacked"),
    "databarstackedcomposite": BarcodeType(
        "databarstackedcomposite", "GS1 DataBar Stacked Composite"
    ),
    "databarstackedomni": BarcodeType(
        "databarstackedomni", "GS1 DataBar Stacked Omnidirectional"
    ),
    "databarstackedomnicomposite": BarcodeType(
        "databarstackedomnicomposite", "GS1 DataBar Stacked Omnidirectional Composite"
    ),
    "databartruncated": BarcodeType("databartruncated", "GS1 DataBar Truncated"),
    "databartruncatedcomposite": BarcodeType(
        "databartruncatedcomposite", "GS1 DataBar Truncated Composite"
    ),
    "datalogic2of5": BarcodeType("datalogic2of5", "Datalogic 2 of 5"),
    "datamatrix": BarcodeType("datamatrix", "Data Matrix"),
    "datamatrixrectangular": BarcodeType(
        "datamatrixrectangular", "Data Matrix Rectangular"
    ),
    "datamatrixrectangularextension": BarcodeType(
        "datamatrixrectangularextension", "Data Matrix Rectangular Extension"
    ),
    "dotcode": BarcodeType("dotcode", "DotCode"),
    "ean13": BarcodeType("ean13", "EAN-13"),
    "ean13composite": BarcodeType("ean13composite", "EAN-13 Composite"),
    "ean14": BarcodeType("ean14", "GS1-14"),
    "ean2": BarcodeType("ean2", "EAN-2 (2 digit addon)"),
    "ean5": BarcodeType("ean5", "EAN-5 (5 digit addon)"),
    "ean8": BarcodeType("ean8", "EAN-8"),
    "ean8composite": BarcodeType("ean8composite", "EAN-8 Composite"),
    "flattermarken": BarcodeType("flattermarken", "Flattermarken"),
    "gs1-128": BarcodeType("gs1-128", "GS1-128"),
    "gs1-128composite": BarcodeType("gs1-128composite", "GS1-128 Composite"),
    "gs1-cc": BarcodeType("gs1-cc", "GS1 Composite 2D Component"),
    "gs1datamatrix": BarcodeType("gs1datamatrix", "GS1 Data Matrix"),
    "gs1datamatrixrectangular": BarcodeType(
        "gs1datamatrixrectangular", "GS1 Data Matrix Rectangular"
    ),
    "gs1dldatamatrix": BarcodeType("gs1dldatamatrix", "GS1 Digital Link Data Matrix"),
    "gs1dlqrcode": BarcodeType("gs1dlqrcode", "GS1 Digital Link QR Code"),
    "gs1dotcode": BarcodeType("gs1dotcode", "GS1 DotCode"),
    "gs1northamericancoupon": BarcodeType(
        "gs1northamericancoupon", "GS1 North American Coupon"
    ),
    "gs1qrcode": BarcodeType("gs1qrcode", "GS1 QR Code"),
    "hanxin": BarcodeType("hanxin", "Han Xin Code"),
    "hibcazteccode": BarcodeType("hibcazteccode", "HIBC Aztec Code"),
    "hibccodablockf": BarcodeType("hibccodablockf", "HIBC Codablock F"),
    "hibccode128": BarcodeType("hibccode128", "HIBC Code 128"),
    "hibccode39": BarcodeType("hibccode39", "HIBC Code 39"),
    "hibcdatamatrix": BarcodeType("hibcdatamatrix", "HIBC Data Matrix"),
    "hibcdatamatrixrectangular": BarcodeType(
        "hibcdatamatrixrectangular", "HIBC Data Matrix Rectangular"
    ),
    "hibcmicropdf417": BarcodeType("hibcmicropdf417", "HIBC MicroPDF417"),
    "hibcpdf417": BarcodeType("hibcpdf417", "HIBC PDF417"),
    "hibcqrcode": BarcodeType("hibcqrcode", "HIBC QR Code"),
    "iata2of5": BarcodeType("iata2of5", "IATA 2 of 5"),
    "identcode": BarcodeType("identcode", "Deutsche Post Identcode"),
    "industrial2of5": BarcodeType("industrial2of5", "Industrial 2 of 5"),
    "interleaved2of5": BarcodeType("interleaved2of5", "Interleaved 2 of 5 (ITF)"),
    "isbn": BarcodeType("isbn", "ISBN"),
    "ismn": BarcodeType("ismn", "ISMN"),
    "issn": BarcodeType("issn", "ISSN"),
    "itf14": BarcodeType("itf14", "ITF-14"),
    "jabcode": BarcodeType("jabcode", "JAB Code (Beta)"),
    "japanpost": BarcodeType("japanpost", "Japan Post 4 State Customer Code"),
    "kix": BarcodeType("kix", "Royal Dutch TPG Post KIX"),
    "leitcode": BarcodeType("leitcode", "Deutsche Post Leitcode"),
    "mailmark": BarcodeType("mailmark", "Royal Mail Mailmark"),
    "mands": BarcodeType("mands", "Marks & Spencer"),
    "matrix2of5": BarcodeType("matrix2of5", "Matrix 2 of 5"),
    "maxicode": BarcodeType("maxicode", "MaxiCode"),
    "micropdf417": BarcodeType("micropdf417", "MicroPDF417"),
    "microqrcode": BarcodeType("microqrcode", "Micro QR Code"),
    "msi": BarcodeType("msi", "MSI Modified Plessey"),
    "onecode": BarcodeType("onecode", "USPS Intelligent Mail"),
    "pdf417": BarcodeType("pdf417", "PDF417"),
    "pdf417compact": BarcodeType("pdf417compact", "Compact PDF417"),
    "pharmacode": BarcodeType("pharmacode", "Pharmaceutical Binary Code"),
    "pharmacode2": BarcodeType("pharmacode2", "Two-track Pharmacode"),
    "planet": BarcodeType("planet", "USPS PLANET"),
    "plessey": BarcodeType("plessey", "Plessey UK"),
    "posicode": BarcodeType("posicode", "PosiCode"),
    "postnet": BarcodeType("postnet", "USPS POSTNET"),
    "pzn": BarcodeType("pzn", "Pharmazentralnummer (PZN)"),
    "qrcode": BarcodeType("qrcode", "QR Code"),
    "rationalizedCodabar": BarcodeType("rationalizedCodabar", "Codabar"),
    "raw": BarcodeType("raw", "Custom 1D symbology"),
    "rectangularmicroqrcode": BarcodeType(
        "rectangularmicroqrcode", "Rectangular Micro QR Code"
    ),
    "royalmail": BarcodeType("royalmail", "Royal Mail 4 State Customer Code"),
    "sscc18": BarcodeType("sscc18", "SSCC-18"),
    "swissqrcode": BarcodeType("swissqrcode", "Swiss QR Code"),
    "symbol": BarcodeType("symbol", "Miscellaneous symbols"),
    "telepen": BarcodeType("telepen", "Telepen"),
    "telepennumeric": BarcodeType("telepennumeric", "Telepen Numeric"),
    "ultracode": BarcodeType("ultracode", "Ultracode"),
    "upca": BarcodeType("upca", "UPC-A"),
    "upcacomposite": BarcodeType("upcacomposite", "UPC-A Composite"),
    "upce": BarcodeType("upce", "UPC-E"),
    "upcecomposite": BarcodeType("upcecomposite", "UPC-E Composite"),
}
